### PR TITLE
Re-implemented togglebookmark off using stack

### PR DESCRIPTION
**Pull request by kkc-dev**
1. Stack is used to know the relation between conditional block inside conditional block
2. If parent is toggle of, no child will have toggle off object

### CODE REVIEW

File: src/__tests__/utils.js [hunk 1]
```diff
@@ -38,6 +38,20 @@ export const getBookmark = (id) => {
 	}]
 }
 
+export const getConditionalBookmark = (id) => {
```
Comment (rek) [inline src/__tests__/utils.js]: this is a generic lib, we don't want to put in our exact use cases, make this more generic, even in the test utils, so function name something like getNamedBookmark, and change COND to something like 'PREFIX::'

Comment (rek) [reply src/__tests__/utils.js]: we can replace the above getBookmark function with this new one. then just use:
```
export const getBookmark = (id, prefix = 'PREFIX::') => {
```

Comment (rek) [reply src/__tests__/utils.js]: don't want to duplicate and have both.

File: src/sfdt/__tests__/fixtures/nestedConditionListWithOneChildToggleOff.js [hunk 2]
```diff
@@ -0,0 +1,70 @@
+export const inlines = [
+  {
+      "bookmarkType": 0,
+      "name": "COND::dafe554d-08b5-463f-a40c-cf5e260be606"
```
Comment (rek) [inline src/sfdt/__tests__/fixtures/nestedConditionListWithOneChildToggleOff.js]: this fixture is pretty short, perhaps we can generated using our helper functions.

File: src/sfdt/processInlines.ts [hunk 3]
```diff
@@ -17,6 +19,21 @@ export default (sfdt, callback) => {
 
 			block.inlines = callback(block.inlines)
 		})
+
+		if (get(section, "headersFooters")) {
+      for (let eachKey in section.headersFooters) {
```
Comment (rek) [inline src/sfdt/processInlines.ts]: spacing is inconsistent. please use the spacing found in this file, eg: tabs

File: src/sfdt/toggleBookmark.ts [hunk 4]
```diff
@@ -5,6 +5,88 @@ import {
 	isBookmarkStart,
 	isBookmarkEnd,
 } from '../queryBookmark'
+import Stack from '../stack'
```
Comment (rek) [inline src/sfdt/toggleBookmark.ts]: i dont think implementing a custom stack adds any benefits. 

perhaps peekAt and pop are helpful to save code, but they can easily be implemented here as functional helpers, rather than a new oop construction.

eg:
```
const pop = (items) => items.length === 0 ? 'Underflow' : items.pop()
```

though i worry about this function having two different return types, that is generally not good practise.

ps. take note to use the Zegal JS style of single quotes, triple equals and always having brackets around function params.

File: src/sfdt/toggleBookmark.ts [hunk 5]
```diff
@@ -5,6 +5,88 @@ import {
...
+
+const makeToggleOff = (inlines, name) => {
```
Comment (rek) [inline src/sfdt/toggleBookmark.ts]: please type all functions

File: src/sfdt/toggleBookmark.ts [hunk 6]
```diff
@@ -5,6 +5,88 @@ import {
...
+  const newInlines = []; // Should act as queue for getting new list of inlines
+  const stackOfBookmarks = new Stack(); // will be used to find relationship of bookmarks
+  let toggleOffAtIndex = null;
+
+  inlines.forEach((inline, index) => {
```
Comment (rek) [inline src/sfdt/toggleBookmark.ts]: note: the previous function had a lot of comments to explain what was going on. please also add comments in your new one to help the next reader. logic blocks like this are hard to follow otherwise

File: src/sfdt/toggleBookmark.ts [hunk 7]
```diff
@@ -5,6 +5,88 @@ import {
...
+    const nextInline = inlines[index + 1];
+    const prevInline = inlines[index - 1];
+
+    if (!(inline.fieldType || inline.hasFieldEnd)) {
+      newInlines.push(inline);
+    }
+
+    if (inline.name && inline.name.split("::").includes("COND")) {
```
Comment (rek) [inline src/sfdt/toggleBookmark.ts]: this function should not know what a 'COND' is. please refactor so we make this a generic toggleBookmark function, not hardcoded to our usecase.

File: src/sfdt/toggleBookmark.ts [hunk 8]
```diff
@@ -5,6 +5,88 @@ import {
...
+const makeToggleOff = (inlines, name) => {
+  const newInlines = []; // Should act as queue for getting new list of inlines
+  const stackOfBookmarks = new Stack(); // will be used to find relationship of bookmarks
+  let toggleOffAtIndex = null;
+
+  inlines.forEach((inline, index) => {
+    const nextInline = inlines[index + 1];
+    const prevInline = inlines[index - 1];
+
+    if (!(inline.fieldType || inline.hasFieldEnd)) {
```
Comment (rek) [inline src/sfdt/toggleBookmark.ts]: sfdt query logic to determine what type of an inilne we are dealing with should live in a separate re-useable function, for an example please see the queryBookmark file.

File: src/sfdt/toggleBookmark.ts [hunk 9]
```diff
@@ -5,6 +5,88 @@ import {
...
+
+        if (isSameBookmark) {
+          if (toggleOffAtIndex === null) {
+            toggleOffAtIndex = stackOfBookmarks.peekAt();
+          } else if (stackOfBookmarks.peekAt() <= toggleOffAtIndex) {
+            toggleOffAtIndex = stackOfBookmarks.peekAt();
+          }
+        }
+
+        if (inline.hasFieldEnd || inline.fieldType === 2) {
```
Comment (rek) [inline src/sfdt/toggleBookmark.ts]: same point here, let's try and remove this type of inline type checking, and have it in a named function instead, eg:

`const isHiddenAreaStartTag = (inline) => ...`
`const isHiddenAreaClosingTag = (inline) => ...`

**Comment by kkc-dev** [comment]
I have updated according to comments besides stack and generating inlines data. @AdamTombleson

File: src/index.ts [hunk 10]
```diff
@@ -39,6 +43,10 @@ const removeBookmarkAndHighlight = safe(unsafe_removeBookmarkAndHighlight)
 const isMatchingBookmark = safe(unsafe_isMatchingBookmark)
 const isBookmarkStart = safe(unsafe_isBookmarkStart)
 const isBookmarkEnd = safe(unsafe_isBookmarkEnd)
+const isToggleStart = safe(unsafe_isToggleStart)
```
Comment (rek) [inline src/index.ts]: the only point of this safe wrapper is to catch documentEditor errors. If your functions don't use the FE documentEditor instance, then you most likely don't need to do this.

Comment (kkc-dev) [reply src/index.ts]: updated @AdamTombleson

**Comment by kkc-dev** [comment]
fix updated to make toggle off in false case. @AdamTombleson

File: src/sfdt/toggleBookmark.ts [hunk 11]
```diff
@@ -131,18 +131,16 @@ const toggleBookmark = (sfdt: any, name: string, toggleOn = true) => {
           }
         }
 
-        if (inMatchingBookmark) {
           // console.log("FOUND, DOING TOGGLE ON");
-          if (inline.fieldType === 2 && isMatchingBookmark(nextInline, name)) {
-            defaultAdd = false
-          }
+        if (inline.fieldType === 2 && isMatchingBookmark(nextInline, name)) {
```
Comment (rek) [inline src/sfdt/toggleBookmark.ts]: as discussed, let's try and use the helper here

**Comment by rek** [comment]
Ready to merge @p-mag
